airflow/operators/sql_branch_operator.py

```diff
@@ -0,0 +1,174 @@
+from airflow.models import BaseOperator, SkipMixin
+from airflow.utils.decorators import apply_defaults
+
+
+class BranchSqlOperator(BaseOperator, SkipMixin):
+    """
+    Executes sql code in a specific database
+
+    :param sql: the sql code to be executed. (templated)
+    :type sql: Can receive a str representing a sql statement or reference to a template file.
+               Template reference are recognized by str ending in '.sql'.
+               Expected SQL query to return Boolean (True/False), integer (0 = False, Otherwise = 1)
+               or string (true/y/yes/1/on/false/n/no/0/off).
+    :param follow_task_ids_if_true: task id or task ids to follow if query return true
+    :type follow_task_ids_if_true: str or list
+    :param follow_task_ids_if_false: task id or task ids to follow if query return true
+    :type follow_task_ids_if_false: str or list
+    :param conn_id: reference to a specific database
+    :type conn_id: str
+    :param database: name of database which overwrite defined one in connection
+    :param parameters: (optional) the parameters to render the SQL query with.
+    :type parameters: mapping or iterable
+    """
+
+    template_fields = ("sql",)
+    template_ext = (".sql",)
+    ui_color = "#a22034"
+    ui_fgcolor = "#F7F7F7"
+
+    @apply_defaults
+    def __init__(
+        self,
+        sql: str,
+        follow_task_ids_if_true: List[str],
+        follow_task_ids_if_false: List[str],
+        conn_id: str = "default_conn_id",
+        database: Optional[str] = None,
+        parameters: Optional[Union[Mapping, Iterable]] = None,
+        *args,
+        **kwargs,
+    ) -> None:
+        super().__init__(*args, **kwargs)
+        self.conn_id = conn_id
+        self.sql = sql
+        self.parameters = parameters
+        self.follow_task_ids_if_true = follow_task_ids_if_true
+        self.follow_task_ids_if_false = follow_task_ids_if_false
+        self.database = database
+        self._hook = None
+
+    def _get_hook(self):
+        self.log.debug("Get connection for %s", self.conn_id)
+        conn = BaseHook.get_connection(self.conn_id)
+
+        allowed_conn_type = {
+            "google_cloud_platform",
+            "jdbc",
+            "mssql",
+            "mysql",
+            "odbc",
+            "oracle",
+            "postgres",
+            "presto",
+            "sqlite",
+            "vertica",
+        }
+        if conn.conn_type not in allowed_conn_type:
+            raise AirflowException(
+                "The connection type is not supported by BranchSqlOperator. "
+                + "Supported connection types: {}".format(list(allowed_conn_type))
+            )
+
+        if not self._hook:
+            self._hook = conn.get_hook()
+            if self.database:
+                self._hook.schema = self.database
+
+        return self._hook
+
+    def execute(self, context: Dict):
+        # get supported hook
+        self._hook = self._get_hook()
+
```

Comment:
Maybe it's a good idea to check if the hook is DbApiHook instance? I think it's even better check than checking the connection types.

Reply:
Hi @potiuk , If I added the following code to check if the hook is DbApiHook instance, my unit test's mock will break. Any suggestion? I searched the Airflow code base but I couldn't find an example to resolve this issue. Thanks!

```Python
if not isinstance(self._hook, DbApiHook):
    raise AirflowException(
        "Unexpected type returned '%s' expected DbApiHook" % type(self._hook))

# In unit test, my test will fail with
"airflow.exceptions.AirflowException: Unexpected type returned '<class 'unittest.mock.MagicMock'>' expected DbApiHook"
```

Reply:
Just rebase it to latest master please. We had some problems with kubernetes tests - they should be largely solved now (I am working on final fix) but at least it should not fail now.

Reply:
Here is some info on that: https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#how-to-rebase-pr

Reply:
Hi @potiuk , I haven't started the breeze integrated test yet. I only run the unit test locally. So, I think this is a "mock" issue.

```Python
pytest ../airflow/tests/operators/test_sql_branch_operator.py
```
In the unit test, I used mock connection (Similar to the existing unit "test test_sql_sensor")
``` Python
mock_hook.get_connection("mysql_default").conn_type = "mysql"
```
Therefore, during my local pytest run, the new "isinstance(self._hook, DbApiHook)" check failed and return error.

```Python
def execute(self, context: Dict):
        # get supported hook
        self._hook = self._get_hook()
    
        if not isinstance(self._hook, DbApiHook):
            raise AirflowException(
>               "Unexpected type returned '%s' expected DbApiHook" % type(self._hook)
            )
E           airflow.exceptions.AirflowException: Unexpected type returned '<class 'unittest.mock.MagicMock'>' expected DbApiHook
```

Reply:
I also checked other operators to see if any of them verify the DbApiHook return type but I couldn't find any example of that check.

Reply:
OK. Fine for me. Indeed that is quite OK this way :)

Reply:
Can you please rebase to latest master? That should fix the failure problem.

Reply:
Thank you so much for reviewing this PR. Yes, I will rebase my PR. Thanks!